protocol/communities/community.go

```diff
@@ -2100,59 +2058,306 @@ func (o *Community) deleteChat(chatID string) *CommunityChanges {
 		o.config.CommunityDescription.TokenPermissions = make(map[string]*protobuf.CommunityTokenPermission)
 	}
-	for _, newPermission := range newPermissions {
-		o.config.CommunityDescription.TokenPermissions[newPermission.Id] = newPermission
+	if _, ok := o.config.CommunityDescription.TokenPermissions[permission.Id]; !ok {
+		return nil, ErrTokenPermissionNotFound
+	}
+
+	changes := o.emptyCommunityChanges()
+	o.config.CommunityDescription.TokenPermissions[permission.Id] = permission
+
+	becomeMemberPermissions := o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER)
+	o.SetEncrypted(len(becomeMemberPermissions) > 0)
+
+	if changes.TokenPermissionsModified == nil {
+		changes.TokenPermissionsModified = make(map[string]*protobuf.CommunityTokenPermission)
 	}
+	changes.TokenPermissionsModified[permission.Id] = o.config.CommunityDescription.TokenPermissions[permission.Id]
+
+	return changes, nil
 }
 
-func (o *Community) addCommunityMember(pk *ecdsa.PublicKey, member *protobuf.CommunityMember) {
+func (o *Community) deleteTokenPermission(permissionID string) (*CommunityChanges, error) {
+	delete(o.config.CommunityDescription.TokenPermissions, permissionID)
+
+	becomeMemberPermissions := o.TokenPermissionsByType(protobuf.CommunityTokenPermission_BECOME_MEMBER)
+	o.SetEncrypted(len(becomeMemberPermissions) > 0)
+
+	changes := o.emptyCommunityChanges()
+	changes.TokenPermissionsRemoved = append(changes.TokenPermissionsRemoved, permissionID)
+	return changes, nil
+}
+
+func (o *Community) addMemberWithRevealedAccounts(memberKey string, roles []protobuf.CommunityMember_Roles, accounts []*protobuf.RevealedAccount) *CommunityChanges {
+	changes := o.emptyCommunityChanges()
 
 	if o.config.CommunityDescription.Members == nil {
 		o.config.CommunityDescription.Members = make(map[string]*protobuf.CommunityMember)
 	}
 
-	memberKey := common.PubkeyToHex(pk)
-	o.config.CommunityDescription.Members[memberKey] = member
+	if _, ok := o.config.CommunityDescription.Members[memberKey]; !ok {
+		o.config.CommunityDescription.Members[memberKey] = &protobuf.CommunityMember{Roles: roles}
+		changes.MembersAdded[memberKey] = o.config.CommunityDescription.Members[memberKey]
+	}
+
+	o.config.CommunityDescription.Members[memberKey].RevealedAccounts = accounts
+	changes.MemberWalletsAdded[memberKey] = o.config.CommunityDescription.Members[memberKey].RevealedAccounts
+
+	return changes
 }
 
-func permissionExists(id string, prevPermissions []*protobuf.CommunityTokenPermission) bool {
-	for _, prevPermission := range prevPermissions {
-		if prevPermission.Id == id {
-			return true
+func (o *Community) DeclineRequestToJoin(dbRequest *RequestToJoin) error {
+	// typically, community's clock is increased implicitly when making changes
+	// to it, however in this scenario there are no changes in the community, yet
+	// we need to increase the clock to ensure the owner event is processed by other
+	// nodes.
```

Comment:
This comment doesn't seem to belong here and should probably be put somewhere around L:2358